vck-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/openid/OpenId4VpVerifier.kt

```diff
@@ -641,70 +645,105 @@ class OpenId4VpVerifier(
+        if (clientId == null || responseUrl == null)
+            throw IllegalStateException("Missing required parameters: clientId, responseUrl")
+
+        val expected = document.calcDeviceAuthenticationOpenId4VpFinal(
+            clientId = clientId,
+            responseUrl = responseUrl,
+            nonce = nonce,
+            encrypted = mdocGeneratedNonce.isNotEmpty()
+        ).wrapAsExpectedPayload()
+
+        verifyCoseSignature(deviceSignature, walletKey, byteArrayOf(), expected).onFailure {
+            val legacy = document.calcDeviceAuthentication(
+                nonce = nonce,
+                mdocGeneratedNonce = mdocGeneratedNonce,
+                clientId = clientId,
+                responseUrl = responseUrl
+            ).wrapAsExpectedPayload()
+            verifyCoseSignature(deviceSignature, walletKey, byteArrayOf(), legacy).onFailure {
+                throw IllegalArgumentException(
+                    "deviceSignature not verified, expected ${expected.encodeToString(Base16())} " +
+                            "or ${legacy.encodeToString(Base16())}", it
+                )
             }
         }
         true
     }
 
+    private fun DeviceAuthentication.wrapAsExpectedPayload(): ByteArray = coseCompliantSerializer
+        .encodeToByteArray(coseCompliantSerializer.encodeToByteArray(this))
+        .wrapInCborTag(24)
+
+    /**
+     * Performs calculation of the [at.asitplus.iso.SessionTranscript] and [at.asitplus.iso.DeviceAuthentication],
+     * acc. to OpenID4VP 1.0
+     */
+    private fun Document.calcDeviceAuthenticationOpenId4VpFinal(
+        clientId: String,
+        responseUrl: String,
+        nonce: String,
+        encrypted: Boolean,
+    ) = DeviceAuthentication(
+        type = DeviceAuthentication.TYPE,
+        sessionTranscript = SessionTranscript.forOpenId(
+            OpenId4VpHandover(
+                type = OpenId4VpHandover.TYPE_OPENID4VP,
+                hash = coseCompliantSerializer.encodeToByteArray<OpenId4VpHandoverInfo>(
+                    OpenId4VpHandoverInfo(
+                        clientId = clientId,
+                        nonce = nonce,
+                        jwkThumbprint = if (encrypted) {
+                            decryptionKeyMaterial.jsonWebKey.sessionTranscriptThumbprint()
+                        } else null,
+                        responseUrl = responseUrl,
+                    )
+                ).sha256(),
+            )
+        ),
+        docType = docType,
+        namespaces = deviceSigned.namespaces
+    )
+
     /**
      * Performs calculation of the [at.asitplus.iso.SessionTranscript] and [at.asitplus.iso.DeviceAuthentication],
      * acc. to ISO/IEC 18013-5:2021 and ISO/IEC 18013-7:2024
      */
+    @Suppress("DEPRECATION")
     private fun Document.calcDeviceAuthentication(
-        challenge: String,
+        nonce: String,
```

Comment:
if you deprecate it and keep it for compatibility, don't rename the param, as it breaks compat

Reply:
It's not deprecated, and it's a private function, i don't see an issue here

Reply:
nvm